core-bundle/src/EventListener/PreviewToolbarListener.php

```diff
@@ -0,0 +1,161 @@
+ */
+final class PreviewToolbarListener
+{
+
+    private $scopeMatcher;
+
+    private $previewScript;
+
+    private $twig;
+
+    private $authorizationChecker;
+
+    private $tokenStorage;
+
+    private $tokenChecker;
+
+    private $router;
+
+    private $tokenManager;
+
+    private $csrfTokenName;
+
+    public function __construct(
+        string $previewScript,
+        ScopeMatcher $scopeMatcher,
+        TwigEnvironment $twig,
+        AuthorizationChecker $authorizationChecker,
+        TokenStorageInterface $tokenStorage,
+        TokenChecker $tokenChecker,
+        RouterInterface $router,
+        CsrfTokenManagerInterface $tokenManager,
+        string $csrfTokenName
+    ) {
+        $this->previewScript        = $previewScript;
+        $this->scopeMatcher         = $scopeMatcher;
+        $this->twig                 = $twig;
+        $this->authorizationChecker = $authorizationChecker;
+        $this->tokenStorage         = $tokenStorage;
+        $this->tokenChecker         = $tokenChecker;
+        $this->router               = $router;
+        $this->tokenManager         = $tokenManager;
+        $this->csrfTokenName        = $csrfTokenName;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        if (!$this->scopeMatcher->isContaoMasterRequest($event)) {
+            return;
+        }
+
+        $request  = $event->getRequest();
+        $response = $event->getResponse();
+
+        if ($request->getScriptName() !== $this->previewScript) {
+            return;
+        }
+
+        // Ignore redirects and errors
+        if (200 !== $response->getStatusCode()) {
+            return;
+        }
+
+        // Do not capture redirects or modify XML HTTP Requests
+        if ($request->isXmlHttpRequest()) {
+            return;
+        }
+
+        if (null === $user = BackendUser::getInstance()) {
+            return;
+        }
+
+        // Only inject toolbar on html responses
+        if ('html' !== $request->getRequestFormat()
+            || false === strpos((string) $response->headers->get('Content-Type'), 'html')
+            || false !== stripos((string) $response->headers->get('Content-Disposition'), 'attachment;')
+        ) {
+            return;
+        }
+
+        try {
+            $this->injectToolbar($response, $request, $user);
+        } catch (TwigError $e) {
+        }
+    }
+
+
+    /**
+     * @throws TwigError
+     */
+    private function injectToolbar(Response $response, Request $request, BackendUser $user): void
+    {
+        $content = $response->getContent();
+
+        if (false === strpos($content, '<body')) {
+            return;
+        }
+
+        $canSwitchUser    = ($user->isAdmin || (!empty($user->amg) && \is_array($user->amg)));
+        $frontendUsername = $this->tokenChecker->getFrontendUsername();
+        $showUnpublished  = $this->tokenChecker->isPreviewMode();
+
+        $toolbar = str_replace(
```

Comment:
Why this `str_replace()`?

Reply:
This removes line-breaks and reduces the injected html,js,css to one line. If you view the page source, it is much more tidy. Symfony Debug Toolbar does this as well :)